File: src/platform/Android/app/src/main/java/com/jappsy/gui/JPaint.java
package com.jappsy.gui;

public class JPaint {
    public float m_scale = 1.0f;
    public boolean m_antialias = false;
    public int m_typeface = JTypeface.DEFAULT;
    public int m_alignX = JAlignX.LEFT;
    public int m_alignY = JAlignY.BASELINE;
    public float m_size = 16.0f;

    public boolean m_invisible = false;
    public int m_color = 0xFF000000;

    public boolean m_strokeInvisible = true;
    public float m_strokeOffset = 0;
    public float m_strokeWidth = 0;
    public int m_strokeColor = 0x00000000;
    
    public boolean m_shadowInvisible = true;
    public int[] m_shadow = {0, 0};
    public int m_shadowWidth = 0;
    public int m_shadowColor = 0x00000000;

    public int m_colorFilter = JPaintColorFilter.PAINT_COLOR_FILTER_NORMAL;
    
    public boolean m_sdfInvisible = true;
    public int[] m_sdfColors = {0x00000000, 0x00000000, 0x00000000, 0x00000000};
    
    public JPaint setSDFColors(int[] colors) {
    	int len = 0;
    	if (colors != null) {
    		len = colors.length;
        	if (len > 4) len = 4;
    	}
    	int i = 0;
    	int t = 0;
    	while (i < len) {
    		m_sdfColors[i] = colors[i];
    		if ((m_sdfColors[i] & 0xFF000000) == 0) t++;
    		i++;
    	}
    	while (i < 4) {
    		m_sdfColors[i] = 0x00000000;
    		t++;
    		i++;
    	}
    	m_sdfInvisible = (t == 4);
    	return this;
    }
    
    public JPaint setAntiAlias(boolean antialias) {
        m_antialias = antialias;
        return this;
    }
    
    public JPaint setTypeface(int typeface) {
        m_typeface = typeface;
        return this;
    }
    
    public JPaint setTextSize(float size) {
        m_size = size;
        return this;
    }
    
    public JPaint setAlignX(int align) {
        m_alignX = align;
        return this;
    }
    
    public JPaint setAlignY(int align) {
        m_alignY = align;
        return this;
    }
    
    public JPaint setColor(int color) {
        m_color = color;
        m_invisible = (m_color & 0xFF000000) == 0;
        return this;
    }
    
    public JPaint setStroke(int width, int color) {
    	m_strokeOffset = -width;
        m_strokeWidth = width;
        m_strokeColor = color;
        m_strokeInvisible = ((color & 0xFF000000) == 0) || (width == 0);
        return this;
    }
    
    public JPaint setStroke(float offset, float width, int color) {
    	m_strokeOffset = offset;
        m_strokeWidth = width;
        m_strokeColor = color;
        m_strokeInvisible = ((color & 0xFF000000) == 0) || (width == 0);
        return this;
    }
    
    public JPaint setAlpha(byte alpha) {
        m_color = (m_color & 0x00FFFFFF) | ((int)alpha << 24);
        m_invisible = (m_color & 0xFF000000) == 0;
        return this;
    }
    
    public JPaint setShadow(int width, int dx, int dy, int color) {
        m_shadow[0] = dx;
        m_shadow[1] = dy;
        m_shadowWidth = width;
        m_shadowColor = color;
        m_shadowInvisible = ((color & 0xFF000000) == 0) || ((dx == 0) && (dy == 0) && (width == 0));
        return this;
    }
    
    public JPaint setScale(float scale) {
        m_scale = scale;
        return this;
    }
    
    public JPaint setColorFilter(int colorFilter) {
        m_colorFilter = colorFilter;
        return this;
    }
}

File: src/__IMPORT/javascript/jappsy/modules/gui/JPaint.js
Jappsy.JPaint = function(paint) {
	if (paint === undefined) {
		this.m_scale = 1.0;
		this.m_antialias = false;
		this.m_typeface = Jappsy.JTypeface.DEFAULT;
		this.m_alignX = Jappsy.JAlignX.LEFT;
		this.m_alignY = Jappsy.JAlignY.BASELINE;
		this.m_size = 16.0;
		
		this.m_invisible = false;
		this.m_color = 0xFF000000;
		this.m_color4iv = new Vec4([0,0,0,1]);
		
		this.m_strokeInvisible = true;
		this.m_strokeOffset = 0;
		this.m_strokeWidth = 0;
		this.m_strokeColor = 0x00000000;
		
		this.m_shadowInvisible = true;
		this.m_shadow = [0, 0];
		this.m_shadowWidth = 0;
		this.m_shadowColor = 0x00000000;
		
		this.m_sdfInvisible = true;
		this.m_sdfColors = [ 0x00000000, 0x00000000, 0x00000000, 0x00000000 ];
	} else {
		this.m_scale = paint.m_scale;
		this.m_antialias = paint.m_antialias;
		this.m_typeface = paint.m_typeface;
		this.m_alignX = paint.m_alignX;
		this.m_alignY = paint.m_alignY;
		this.m_size = paint.m_size;
		
		this.m_invisible = paint.m_invisible;
		this.m_color = paint.m_color;
		this.m_color4iv = new Vec4(paint.m_color4iv);
		
		this.m_strokeInvisible = paint.m_strokeInvisible;
		this.m_strokeOffset = paint.m_strokeOffset;
		this.m_strokeWidth = paint.m_strokeWidth;
		this.m_strokeColor = paint.m_strokeColor;
		
		this.m_shadowInvisible = paint.m_shadowInvisible;
		this.m_shadow = paint.m_shadow.slice(0);
		this.m_shadowWidth = paint.m_shadowWidth;
		this.m_shadowColor = paint.m_shadowColor;
		
		this.m_sdfInvisible = paint.m_sdfInvisible;
		this.m_sdfColors = paint.m_sdfColors.slice(0);
	}
};

Jappsy.JPaint.prototype.setSDFColors = function (colors) {
	var count = colors.length;
	if (count > 4) count = 4;
	for (var i = 0; i < count; i++)
		this.m_sdfColors[i] = colors[i];
	for (var i = count; i < 4; i++)
		this.m_sdfColors[i] = 0x00000000;
	this.m_sdfInvisible = (count == 0);
	return this;
};

Jappsy.JPaint.prototype.setAntiAlias = function (antialias) {
	this.m_antialias = antialias;
	return this;
};

Jappsy.JPaint.prototype.setTypeface = function (typeface) {
	this.m_typeface = typeface;
	return this;
};

Jappsy.JPaint.prototype.setTextSize = function (size) {
	this.m_size = size;
	return this;
};

Jappsy.JPaint.prototype.setAlignX = function (align) {
	this.m_alignX = align;
	return this;
};

Jappsy.JPaint.prototype.setAlignY = function (align) {
	this.m_alignY = align;
	return this;
};

Jappsy.JPaint.prototype.setColor = function (color) {
	this.m_color = color;
	this.m_invisible = (color & 0xFF000000) == 0;
	this.m_color4iv[0] = ((color >>> 16) & 0xFF) / 255.0;		// R
	this.m_color4iv[1] = ((color >>> 8) & 0xFF) / 255.0;		// G
	this.m_color4iv[2] = (color & 0xFF) / 255.0;				// B
	this.m_color4iv[3] = ((color >>> 24) & 0xFF) / 255.0;		// A
	return this;
};

Jappsy.JPaint.prototype.setLight = function (light, alpha) {
	this.m_color4iv[0] = this.m_color4iv[1] = this.m_color4iv[2] = light;
	this.m_color4iv[3] = alpha;
	this.m_color =
		Math.floor(Jappsy.clamp(this.m_color4iv[2] * 255.0, 0, 255)) |
		(Math.floor(Jappsy.clamp(this.m_color4iv[1] * 255.0, 0, 255)) << 8) |
		(Math.floor(Jappsy.clamp(this.m_color4iv[0] * 255.0, 0, 255)) << 16) |
		(Math.floor(Jappsy.clamp(this.m_color4iv[3] * 255.0, 0, 255)) << 24);
	this.m_invisible = (this.m_color & 0xFF000000) == 0;
	return this;
};

Jappsy.JPaint.prototype.setStroke = function (width, color) {
	this.m_strokeOffset = -width;
	this.m_strokeWidth = width;
	this.m_strokeColor = color;
	this.m_strokeInvisible = ((color & 0xFF000000) == 0) || (width == 0);
	return this;
};

Jappsy.JPaint.prototype.setAlpha = function (alpha) {
	this.m_color = (this.m_color & 0x00FFFFFF) | ((alpha & 0xFF) << 24);
	this.m_invisible = (this.m_color & 0xFF000000) == 0;
	return this;
};

Jappsy.JPaint.prototype.setShadow = function (width, dx, dy, color) {
	this.m_shadow = [ dx, dy ];
	this.m_shadowWidth = width;
	this.m_shadowColor = color;
	this.m_shadowInvisible = ((color & 0xFF000000) == 0) || ((dx == 0) && (dy == 0) && (width == 0));
	return this;
};

Jappsy.JPaint.prototype.setScale = function (scale) {
	this.m_scale = scale;
	return this;
};
